operator/pkg/tasks/init/crd.go

```diff
@@ -73,7 +80,17 @@ func skipCrdsDownload(r workflow.RunData) (bool, error) {
 		return false, errors.New("prepare-crds task invoked with an invalid data struct")
 	}
 
-	crdsDir := path.Join(data.DataDir(), data.KarmadaVersion())
+	crdTarball := data.CrdTarball()
+	if crdTarball.CRDDownloadPolicy != nil && *crdTarball.CRDDownloadPolicy == operatorv1alpha1.DownloadAlways {
+		klog.V(2).InfoS("[skipCrdsDownload] CrdDownloadPolicy is 'Always', skipping cache check")
+		return false, nil
+	}
+
```

Comment:
1. This code is not compatible with the previous behavior.

2. I believe it is necessary to be compatible with the previous behavior, as this will facilitate user upgrades.

Therefore, I suggest that this code needs to check whether the path `path.Join(data.DataDir(), data.KarmadaVersion())` exists. If it does, it should also skip the CRD download."

Reply:
As I mentioned in your proposal earlier, we have already stored the CRD files in this path in advance. We don't want the karmada-operator to download any files from the internet.

Reply:
I understand your concern is about changing the cache directory, before this feature, the tarball is cached at `data.DataDir()/<KarmadaVersion>`, and after this feature, the path becomes `data.DataDir()/cache/<hashkey>`. 

Since you embed the tarball into the image, after this change you have to adopt the new path, which probably needs to make some changes to your Dockerfile. Am I right?

Reply:
>  which probably needs to make some changes to your Dockerfile

Yes, I need to modify the Dockerfile. But I am more concerned about: I am not clear about the value of `hashkey` in the path `data.DataDir()/cache/<hashkey>`. It seems that every Karmada version needs to guess what this value is.

Reply:
The approach you're using for pre-loading the CRD tarball can still work. For instance,  the CRD tarball for version `v1.10.3`, by default, would be downloaded via this URL:  `https://github.com/karmada-io/karmada/releases/download/v1.10.3/crds.tar.gz`. To preload, just store it at location `data.DataDir()/cache/<cacheKey>`,  where `cacheKey` is simply the SHA-256 hash of that URL. It's deterministic.

Reply:
On the one hand, the cache maintained by the karmada-operator doesn't guarantee compatibility over versions, as during the upgrading, the cache will be dropped along with the new container. That means we can change the cache directory freely without concern about compatibility. 
@chaunceyjiang In your case, you don't want the karmada-operator to download anything from a remote server, I think we can consider enhancing it by providing a solution to support that case.

On the other hand, we can revisit the reason why introduce the hash-key to the cache directory.

Echo from the proposal [Custom CRD Download Strategy Support for Karmada Operator](https://github.com/karmada-io/karmada/tree/master/docs/proposals/operator-custom-crd-download-strategy):
> Currently, the cache for previously downloaded CRDs lives on the local file system at /var/lib/karmada/<operator-release-version>. To support custom CRD download strategies, we need to a new directory structure for the cache. The Karmada operator can potentially be managing the lifecycle of multiple Karmada instances of heterogeneous versions, and as such, may have to download CRD tarballs from multiple sources. We need to cache the CRDs for each source. Given some source, it's cache entry will live on the local file system at path /var/lib/karmada/cache/<cache-key>. The cache key will be the computed SHA-256 hash of some property derived from that source. In the case of HTTPSource, it will be the hash of the URL.

The reason why introducing the hash-key is to identify different source, my question is do we really need to do that? As the tarball should be consistent even it comes from different sources. Is there some thing I don't get here?

Reply:
I think this caching strategy works just fine and follows an approach similar to how OCI images are pulled from registries. 

With this feature, although the directory structure will change to accommodate the use of custom download strategies, the use case for pre-populating the CRD in the filesystem can still easily be supported.

However, given that this design is extensible, I do think we can provide a more natural solution for this use case. Today, an `HTTPSource` is supported, but, for instance, it's possible to also support a `LocalFileSystemSource` that may help solve this problem more naturally, but I think that's out of scope for this PR.

Reply:
https://github.com/calvin0327/karmada/blob/master/cluster/images/operator-offline.Dockerfile#L22-L27

This is the Dockerfile we are using. How do I calculate the `cacheKey`? Can I use `sha256sum`?

Reply:
Yes.

Reply:
@chaunceyjiang , any other comments/concerns?

Reply:
Thank you for your reply.